anvil/layers.py

```diff
@@ -0,0 +1,1188 @@
+    associated with the probability masses given by the heights.
+
+    The inverse coupling transformation is
+
+        \phi = C^{-1}(x, {p_k}) = \phi_{k-1} + \alpha p_k
+
+    where \phi_{k-1} = \sum_{k'=1}^{k-1} p_{k'} is the (k-1)-th knot point, and \alpha is the
+    fractional position of x in the k-th bin, which is (x - (k-1) * w) / w.
+
+    The gradient of the forward transformation is simply
+
+        dx / d\phi = w / p_k
+
+    Parameters
+    ----------
+    size_half: int
+        Half of the configuration size, which is the size of the input vector for the
+        neural network.
+    hidden_shape: list
+        list containing hidden vector sizes the neural network.
+    activation: str
+        string which is a key for an activation function for all but the final layers
+        of the network.
+    even_sites: bool
+        dictates which half of the data is transformed as a and b, since successive
+        affine transformations alternate which half of the data is passed through
+        neural network.
+
+    Attributes
+    ----------
+    network: torch.nn.Module
+        the dense layers of network h, values are intialised as per the default
+        initialisation of `nn.Linear`
+
+    Methods
+    -------
+    forward(x_input, log_density)
+        see docstring for anvil.layers
+    """
+
+    def __init__(
+        self,
+        size_half: int,
+        *,
+        n_segments: int,
+        hidden_shape: list,
+        activation: str,
+        even_sites: bool,
+    ):
+        super().__init__(size_half, even_sites)
+        self.size_half = size_half
+        self.n_segments = n_segments
+        self.width = 1 / n_segments
+
+        eps = 1e-6  # prevent rounding error which causes sorting into -1th bin
+        self.x_knot_points = torch.linspace(-eps, 1 + eps, n_segments + 1).view(1, -1)
+
+        self.network = NeuralNetwork(
+            size_in=size_half,
+            size_out=size_half * n_segments,
+            hidden_shape=hidden_shape,
+            activation=activation,
+            final_activation=activation,
+        )
+        self.norm_func = nn.Softmax(dim=2)
+
+    def forward(self, x_input, log_density, *unused):
+        """Forward pass of the linear spline layer."""
+        x_a = x_input[:, self._a_ind]
+        x_b = x_input[:, self._b_ind]
+
+        net_out = self.norm_func(
+            self.network(x_a - 0.5).view(-1, self.size_half, self.n_segments)
+        )
+        phi_knot_points = torch.cat(
+            (
+                torch.zeros(net_out.shape[0], self.size_half, 1),
+                torch.cumsum(net_out, dim=2),
+            ),
+            dim=2,
+        )
+
+        # Sort x_b into the appropriate bin
+        # NOTE: need to make x_b contiguous, otherwise searchsorted returns nonsense
+        k_ind = searchsorted(self.x_knot_points, x_b.contiguous()) - 1
+        k_ind.unsqueeze_(dim=-1)
```

Comment:
I was looking at the official searchsorted and I think it really does the same thing and wouldn't require calling `contiguous`. Should we swap this now?

Reply:
yes absolutely